### PR TITLE
Build up specificity for nested routes.

### LINKED ISSUE #102: Nested Route Specificity
I did some debugging on an issue I ran into in the Ember app I work on (v2.6.0) after updating the route-recognizer past 0.1.5, in which the specificity of some routes changed. It appears that the specificity of an array of nested routes is only given by the last one.

Here's a test case that passes before and fails after 0499446f80c649a2df64ced5cc5458742a57c42:

```
test("Matches the route with the longer static prefix.", function() {
  var handler1 = { handler: 1 };
  var handler2 = { handler: 2 };
  var router = new RouteRecognizer();

  router.add([{ path: "/:dynamic", handler: handler1 }, { path: "/", handler: handler1 }]);
  router.add([{ path: "/static", handler: handler2 }, { path: "/", handler: handler2 }]);

  resultsMatch(router.recognize("/static"), [
    { handler: handler2, params: { }, isDynamic: false },
    { handler: handler2, params: { }, isDynamic: false }
  ]);
});
```

Is this an expected change, a regression, or something that was never specified? Would love to hear folks' thoughts.

### PR DESCRIPTION
**Pull request by tjni**
When adding an array of routes, each route will increase the
specificity instead of using only that of the last route.

Resolves #102.

### CODE REVIEW

**Comment by mixonic** [comment]
@tjni you are on the right track here. Specificity of a previously added route was being stomped by a second route, and using an object as a bucket avoids the stomping. However I'm not sure this restores all the functionality we need (there are definitely two tests I have added that do not pass with it) however this is a step in the right direction.

**Comment by mixonic** [comment]
@tjni this seems an incremental fix, but fails on the following assertion:

```js
test("Matches the route with the longer static prefix with nesting", function() {
  var handler1 = { handler: 1 };
  var handler2 = { handler: 2 };
  var router = new RouteRecognizer();

  router.add([{ path: "/", handler: handler1 }, { path: ":post_id", handler: handler1 }]);
  router.add([{ path: "/team", handler: handler2 }, { path: ":user_slug", handler: handler2 }]);

  resultsMatch(router.recognize("/team"), [
    { handler: handler2, params: { }, isDynamic: false },
    { handler: handler2, params: { }, isDynamic: false }
  ]);
});
```

Which I believe is appropriate to assert https://github.com/emberjs/ember.js/issues/13960 is fixed.

**Comment by tjni** [comment]
Planning on looking into this in a few hour in the evening, unless someone else is already.

**Comment by mixonic** [comment]
@tjni in fact the test you've added here already passes on master, I believe.

**Comment by mixonic** [comment]
scratch that, my test does not pass in 0.1.5 either. I must have incorrect ported https://github.com/emberjs/ember.js/issues/13960

**Comment by mixonic** [comment]
@tjni I've confirmed you fix with this test: https://github.com/tildeio/route-recognizer/pull/105/files#diff-d4ce617de337dd77d23737176f335d8cR884 Please add it to this PR! I believe that will give us the option for moving forward with this change instead of reverting.

**Comment by nathanhammond** [comment]
LGTM

File: lib/route-recognizer.js [hunk 1]
```diff
@@ -151,8 +148,6 @@ function parse(route, names, specificity, shouldDecodes) {
     }
   }
 
-  specificity.val = +specificity.val;
```
Comment (nathanhammond) [inline lib/route-recognizer.js]: This is an unnecessary type conversion.

File: lib/route-recognizer.js [hunk 2]
```diff
@@ -125,10 +125,7 @@ function parse(route, names, specificity, shouldDecodes) {
   // `x`, irrespective of the other parts.
   // Because of this similarity, we assign each type of segment a number value written as a
   // string. We can find the specificity of compound routes by concatenating these strings
-  // together, from left to right. After we have looped through all of the segments,
-  // we convert the string to a number.
-  specificity.val = '';
```
Comment (nathanhammond) [inline lib/route-recognizer.js]: Moved to https://github.com/tildeio/route-recognizer/pull/103/files#diff-3cd8d4dfb826ed2ebdceb9c43e729d8eR335

File: lib/route-recognizer.js [hunk 3]
```diff
@@ -249,7 +244,7 @@ State.prototype = {
 // Sort the routes by specificity
 function sortSolutions(states) {
   return states.sort(function(a, b) {
-    return b.specificity.val - a.specificity.val;
+    return (b.specificity.val < a.specificity.val) ? -1 : (b.specificity.val === a.specificity.val) ? 0 : 1;
```
Comment (nathanhammond) [inline lib/route-recognizer.js]: This modification leaves it as a string comparison which gives us the behavior that we want.
[See section 11.8.5.4 of the ECMAScript specification.](http://www.ecma-international.org/ecma-262/5.1/#sec-11.8.5)

**Comment by nathanhammond** [comment]
@mixonic It turns out we _did_ need to land the test from #85.

I've extracted exactly the routes generated from https://github.com/emberjs/ember.js/issues/14073 and we trigger the same failure as the test from #85.

Here is what Ember generates to feed into route-recognizer:
```
  router.add([{"path":"/application_loading","handler":"application_loading"}], {"as":"application_loading"});
  router.add([{"path":"/_unused_dummy_error_path_route_application/:error","handler":"application_error"}], {"as":"application_error"});
  router.add([{"path":"/","handler":"application"},{"path":"/loading","handler":"loading"}], {"as":"loading"});
  router.add([{"path":"/","handler":"application"},{"path":"/_unused_dummy_error_path_route_application/:error","handler":"error"}], {"as":"error"});
  router.add([{"path":"/","handler":"application"},{"path":"/test1_loading","handler":"test1_loading"}], {"as":"test1_loading"});
  router.add([{"path":"/","handler":"application"},{"path":"/_unused_dummy_error_path_route_test1/:error","handler":"test1_error"}], {"as":"test1_error"});
  router.add([{"path":"/","handler":"application"},{"path":"/","handler":"test1"},{"path":"/loading","handler":"test1.loading"}], {"as":"test1.loading"});
  router.add([{"path":"/","handler":"application"},{"path":"/","handler":"test1"},{"path":"/_unused_dummy_error_path_route_test1/:error","handler":"test1.error"}], {"as":"test1.error"});
  router.add([{"path":"/","handler":"application"},{"path":"/","handler":"test1"},{"path":"/test2_loading","handler":"test1.test2_loading"}], {"as":"test1.test2_loading"});
  router.add([{"path":"/","handler":"application"},{"path":"/","handler":"test1"},{"path":"/_unused_dummy_error_path_route_test2/:error","handler":"test1.test2_error"}], {"as":"test1.test2_error"});
  router.add([{"path":"/","handler":"application"},{"path":"/","handler":"test1"},{"path":"/test2","handler":"test1.test2"}], {"as":"test1.test2"});
  router.add([{"path":"/","handler":"application"},{"path":"/","handler":"test1"},{"path":"/index_loading","handler":"test1.index_loading"}], {"as":"test1.index_loading"});
  router.add([{"path":"/","handler":"application"},{"path":"/","handler":"test1"},{"path":"/_unused_dummy_error_path_route_index/:error","handler":"test1.index_error"}], {"as":"test1.index_error"});
  router.add([{"path":"/","handler":"application"},{"path":"/","handler":"test1"},{"path":"/","handler":"test1.index"}], {"as":"test1.index"});
  router.add([{"path":"/","handler":"application"},{"path":"/","handler":"test1"},{"path":"/","handler":"test1.index"}], {"as":"test1"});
  router.add([{"path":"/","handler":"application"},{"path":"/","handler":"test1"},{"path":"/","handler":"test1.index"}], {"as":"application"});
  router.add([{"path":"/","handler":"application"},{"path":"/misc_loading","handler":"misc_loading"}], {"as":"misc_loading"});
  router.add([{"path":"/","handler":"application"},{"path":"/_unused_dummy_error_path_route_misc/:error","handler":"misc_error"}], {"as":"misc_error"});
  router.add([{"path":"/","handler":"application"},{"path":"/:param","handler":"misc"}], {"as":"misc"});
```

**Pull request merged**